### PR TITLE
Add timeline plot with plotly as backend

### PR DESCRIPTION
**Pull request by eukaryo**
## Motivation

To provide an easy way to check the execution status, I want to implement a function to plot the Trials as a timeline.
The timeline here is a Gantt chart with time on the horizontal axis and trial number on the vertical axis, which makes it clear at a glance whether trials are being executed in parallel and whether there are any trials with abnormally long execution times.

## Description of the changes

- Add timeline plot using plotly's horizontal bar graph in `optuna/visualization/_timeline.py`.

- Add Unit tests in `tests/visualization_tests/test_timeline.py`.

test-fig-plotly

### CODE REVIEW

**Comment by eukaryo** [comment]
I see. I have added a figure of timeline plot to the description above. Here is sample code to generate the figure, which is basically the same code as in the docstring of `optuna.visualization.plot_timeline`.
```
import time
import plotly
import optuna

def objective(trial):
    x = trial.suggest_float("x", 0, 1)
    time.sleep(x * 0.1)
    if x > 0.8:
        raise ValueError()
    if x > 0.4:
        raise optuna.TrialPruned()
    return x**2

study = optuna.create_study(direction="minimize")
study.enqueue_trial({"x": 0.3})
study.enqueue_trial({"x": 0.5})
study.enqueue_trial({"x": 0.9})
study.optimize(objective, n_trials=50, n_jobs=2, catch=(ValueError,))

ax = optuna.visualization.plot_timeline(study)
plotly.io.write_image(ax, "test-fig-plotly.png", format="png")
```

File: optuna/visualization/_timeline.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
+import datetime
+from typing import List
+from typing import NamedTuple
+
+from optuna._experimental import experimental_func
+from optuna.logging import get_logger
+from optuna.study import Study
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+from optuna.visualization._pareto_front import _make_hovertext
```
Comment (HideakiImamura) [inline optuna/visualization/_timeline.py:10]: It is not good practice to rely on private functions for other visualization functions. If you want to use this function, why not move it to `optuna.visualization._utils`?

File: optuna/visualization/_timeline.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+_logger = get_logger(__name__)
+
+
+class _TimelineBarInfo(NamedTuple):
+    number: int
+    start: datetime.datetime
+    end: datetime.datetime
+    state: TrialState
+    frozen_trial: FrozenTrial
```
Comment (HideakiImamura) [inline optuna/visualization/_timeline.py:25]: his class, only this `frozen_trial` attribute seems too granular. For example, why not read `_make_hovertext` first before creating this class and have that text as an attribute of str?

File: tests/visualization_tests/test_timeline.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+from optuna.visualization._plotly_imports import go
+import optuna.visualization._timeline
+from optuna.visualization._timeline import _get_timeline_info
+
+
+@pytest.mark.parametrize(
+    "n",
+    [0, 3],
+)
+def test_get_timeline_info(n: int) -> None:
```
Comment (HideakiImamura) [inline tests/visualization_tests/test_timeline.py:19]: How about testing the case when
(1) `trial.datetime_start` is None
(2) `trial.datetime_complete` is None
(3) Both of above

File: tests/visualization_tests/test_timeline.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    for _ in range(n):
+        study.add_trial(
+            optuna.trial.create_trial(
+                params={"x": 0.5},
+                distributions={"x": optuna.distributions.FloatDistribution(0, 1)},
+                value=0,
+                state=optuna.trial.TrialState.COMPLETE,
+            )
+        )
+    assert len(_get_timeline_info(study).bars) == n
```
Comment (HideakiImamura) [inline tests/visualization_tests/test_timeline.py:30]: Might be good to assert the `_TimelineInfo` itself.

Comment (HideakiImamura) [reply tests/visualization_tests/test_timeline.py:30]: Please see [other visualization tests](https://github.com/optuna/optuna/blob/master/tests/visualization_tests/test_contour.py#L207) for example.

File: tests/visualization_tests/test_timeline.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    if n > 0:
+        assert type(_get_timeline_info(study).bars[0].frozen_trial) is FrozenTrial
+
+
+def _create_study(xs: List[TrialState]) -> Study:
```
Comment (HideakiImamura) [inline tests/visualization_tests/test_timeline.py:35]: Might be good to use more descriptive variable name, e.g. `xs` -> `trial_states_list`.

File: tests/visualization_tests/test_timeline.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    study = optuna.create_study()
+    for i, x in enumerate(xs):
+        study.add_trial(
+            optuna.trial.create_trial(
+                params={"x": float(i)},
+                distributions={"x": optuna.distributions.FloatDistribution(-1.0, float(len(xs)))},
```
Comment (HideakiImamura) [inline tests/visualization_tests/test_timeline.py:41]: I'm not sure but it is not good practice to consider `TrialState` as `str`.

File: tests/visualization_tests/test_timeline.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    [
+        [TrialState.COMPLETE, TrialState.PRUNED, TrialState.FAIL],
+        [TrialState.FAIL, TrialState.PRUNED, TrialState.COMPLETE],
+    ],
+)
+def test_get_timeline_plot(xs: List[TrialState]) -> None:
+    study = _create_study(xs)
+    info = _get_timeline_info(study)
+
+    fig = optuna.visualization._timeline._get_timeline_plot(info)
```
Comment (HideakiImamura) [inline tests/visualization_tests/test_timeline.py:60]: Can be simplified with the public API.
```suggestion
    fig = plot_timeline(study)
```

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR. I have several comments. Please take a look.

File: tests/visualization_tests/test_timeline.py (hunk 7 above)
Comment (eukaryo) [reply tests/visualization_tests/test_timeline.py]: Agreed. I revised the code in https://github.com/optuna/optuna/pull/4470/commits/d4e07207df3079fa308987057d6df2dc7ff29492 .

File: tests/visualization_tests/test_timeline.py (hunk 6 above)
Comment (eukaryo) [reply tests/visualization_tests/test_timeline.py]: Agreed; actually, I think I do not consider `TrialState` as `str` in `_create_study` function.

File: tests/visualization_tests/test_timeline.py (hunk 5 above)
Comment (eukaryo) [reply tests/visualization_tests/test_timeline.py]: Agreed. I revised the code in https://github.com/optuna/optuna/pull/4470/commits/d4e07207df3079fa308987057d6df2dc7ff29492 .

File: tests/visualization_tests/test_timeline.py (hunk 4 above)
Comment (eukaryo) [reply tests/visualization_tests/test_timeline.py]: I agree with the principle that we should follow the existing style of tests. In this particular case, however, I don't think it is not a good way to pre-write down and test the exact same intermediate data. The reason is that `study.add_trial` gets and puts in the time information on the fly.

File: tests/visualization_tests/test_timeline.py (hunk 3 above)
Comment (eukaryo) [reply tests/visualization_tests/test_timeline.py:18]: Agreed. I revised the code in https://github.com/optuna/optuna/commit/d4e07207df3079fa308987057d6df2dc7ff29492 . 
According to the implementation of [`create_trial`](https://github.com/optuna/optuna/blob/master/optuna/trial/_frozen.py#L558-L566) , there is no `TrialState` such that `trial.datetime_start` is None and `trial.datetime_complete` is not None.

File: optuna/visualization/_timeline.py (hunk 2 above)
Comment (eukaryo) [reply optuna/visualization/_timeline.py]: Agreed. I revised the code in https://github.com/optuna/optuna/pull/4470/commits/beecdca130691b50ee2bd1a2deeb38432005f98e .

File: optuna/visualization/_timeline.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+from optuna.logging import get_logger
+from optuna.study import Study
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+from optuna.visualization._pareto_front import _make_hovertext
+from optuna.visualization._plotly_imports import _imports
+
+
+if _imports.is_successful():
+    from optuna.visualization._plotly_imports import go  # px, go
```
[the remark this answers is not in the document]
Comment (eukaryo) [reply optuna/visualization/_timeline.py]: I removed the comment.

File: optuna/visualization/_timeline.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+
+
+class _TimelineInfo(NamedTuple):
+    bars: List[_TimelineBarInfo]
+
+
+@experimental_func("3.2.0")
+def plot_timeline(
+    study: Study,
+) -> "go.Figure":
```
[the remark this answers is not in the document]
Comment (eukaryo) [reply optuna/visualization/_timeline.py]: Thank you. I revised the code.

File: optuna/visualization/_timeline.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+                    raise ValueError()
+                if x > 0.4:
+                    raise optuna.TrialPruned()
+                return x ** 2
+
+
+            study = optuna.create_study(direction="minimize")
+            study.optimize(
+                objective, n_trials=50, n_jobs=2, catch=(ValueError,)
+            )
```
[the remark this answers is not in the document]
Comment (eukaryo) [reply optuna/visualization/_timeline.py:57]: I think `n_jobs=2` is essential to show the usefulness of the timeline plot. Timeline plot can visualize trials with overlapping execution time (e.g., in distributed environments).

File: optuna/visualization/_timeline.py [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+
+            fig = optuna.visualization.plot_timeline(study)
+            fig.show()
+
+    Args:
+        study:
+            A :class:`~optuna.study.Study` object whose trials are plotted for
+            their objective value.
```
[the remark this answers is not in the document]
Comment (eukaryo) [reply optuna/visualization/_timeline.py]: I see. I revised the code.

File: optuna/visualization/_timeline.py (hunk 1 above)
Comment (eukaryo) [reply optuna/visualization/_timeline.py]: Agreed. I moved `_make_hovertext` to `optuna.visualization._utils`. This required some additional changes of existing source code.

File: optuna/visualization/_utils.py [hunk 12]
```diff
@@ -179,3 +180,27 @@ def _target(t: FrozenTrial) -> float:
...
+def _make_json_compatible(value: Any) -> Any:
+    try:
+        json.dumps(value)
+        return value
+    except TypeError:
+        # The value can't be converted to JSON directly, so return a string representation.
+        return str(value)
+
+
+def _make_hovertext(trial: FrozenTrial) -> str:
```
Comment (toshihikoyanase) [inline optuna/visualization/_utils.py:194]: I'm not fully sure, but how about removing the `_` prefix and exposing this function?
Currently, only `is_available` is included in `__all__`, but `_make_hovertext` is also used in the other files such as `_pareto_front.py` and `_timeline.py`.

CC @HideakiImamura 

```suggestion
def make_hovertext(trial: FrozenTrial) -> str:
```

**Review by toshihikoyanase — COMMENTED** [review]
Thank you for your PR and example. Let me share some comments on the implementation of the timeline plot.
Next, I'll check the test cases.

File: tests/visualization_tests/test_timeline.py [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+from optuna.study.study import Study
+from optuna.trial import TrialState
+from optuna.visualization._plotly_imports import go
+from optuna.visualization._timeline import _get_timeline_info
+from optuna.visualization._timeline import plot_timeline
+
+
+@pytest.mark.parametrize(
+    "n",
+    [0, 3],
```
Comment (toshihikoyanase) [inline tests/visualization_tests/test_timeline.py:16]: In the other test files, we create a separate test function for the case of an empty study. How about creating `test_get_timeline_info_empty`?

File: tests/visualization_tests/test_timeline.py [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+    info = _get_timeline_info(study)
+    assert len(info.bars) == n
+    if n == 3:
+        trials = study.get_trials()
+        assert trials[0].datetime_start is not None
+        assert trials[1].datetime_start is not None
+        assert trials[2].datetime_start is None
+        assert trials[0].datetime_complete is not None
+        assert trials[1].datetime_complete is None
+        assert trials[2].datetime_complete is None
```
Comment (toshihikoyanase) [inline tests/visualization_tests/test_timeline.py:39]: These assertions are not for the test of timeline info. Could you remove them to keep the test simple?

File: tests/visualization_tests/test_timeline.py [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+        for x in info.bars:
+            assert type(x.hovertext) is str
+            assert x.start is not None
+            assert x.end is not None
+            assert x.start <= x.end
```
Comment (toshihikoyanase) [inline tests/visualization_tests/test_timeline.py:44]: I have some comments on this block:

- Could you rename `x` to `bar` to improve the readability?
- Could you check `bar.number` and `bar.state` as well?
- Could you check the type of `bar.start` and `bar.end`? The current test case accepts numbers, but the plotly function expects `datetime.datetime` for them.

```suggestion
        for bar, trial in zip(info.bars, study.get_trials()):
            assert bar.number == trial.number
            assert bar.state == trial.state
            assert type(bar.hovertext) is str
            assert isinstance(bar.start, datetime.datetime)
            assert isinstance(bar.end, datetime.datetime)
            assert bar.start <= bar.end
```

File: tests/visualization_tests/test_timeline.py [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+    study = optuna.create_study()
+    for i in range(n):
+        study.add_trial(
+            optuna.trial.create_trial(
+                params={"x": 0.5},
+                distributions={"x": optuna.distributions.FloatDistribution(0, 1)},
+                value=0,
+                state=states[i],
+            )
+        )
```
Comment (toshihikoyanase) [inline tests/visualization_tests/test_timeline.py:29]: Maybe we can use `_create_study` to simplify the test.

**Review by toshihikoyanase — COMMENTED** [review]
Let me share my comments on the tests.

File: optuna/visualization/_pareto_front.py [hunk 17]
```diff
@@ -413,22 +404,7 @@ def _make_scatter_object(
...
-        {
-            "number": trial.number,
-            "values": trial.values,
-            "params": trial.params,
-            **user_attrs_dict,
-        },
-        indent=2,
-    )
-    return text.replace("\n", "<br>")
+        raise AssertionError("Must not reach here")
```
[the remark this answers is not in the document]
Comment (eukaryo) [reply optuna/visualization/_pareto_front.py]: I see. I reverted it.

File: optuna/visualization/_timeline.py [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+    hovertext: str
+
+
+class _TimelineInfo(NamedTuple):
+    bars: List[_TimelineBarInfo]
+
+
+@experimental_func("3.2.0")
+def plot_timeline(study: Study) -> "go.Figure":
+    """Plot the timeline of a study.
```
[the remark this answers is not in the document]
Comment (eukaryo) [reply optuna/visualization/_timeline.py:33]: I added the line.

File: optuna/visualization/_timeline.py [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+            go.Bar(
+                name=s.name,
+                x=[(b.end - b.start).total_seconds() * 1000 for b in bars],
+                y=[b.number for b in bars],
+                base=[b.start.isoformat() for b in bars],
+                text=[b.hovertext for b in bars],
+                hovertemplate="%{text}<extra>" + s.name + "</extra>",
+                orientation="h",
+                marker=dict(color=_cm[s.name]),
+                textposition="none",  # avoid drawing hovertext in a bar
```
[the remark this answers is not in the document]
Comment (eukaryo) [reply optuna/visualization/_timeline.py]: I see. I revised the code.

File: tests/visualization_tests/test_timeline.py (hunk 13 above)
Comment (eukaryo) [reply tests/visualization_tests/test_timeline.py]: Agreed. I divided the cases into separate functions.

File: tests/visualization_tests/test_timeline.py (hunk 14 above)
Comment (eukaryo) [reply tests/visualization_tests/test_timeline.py]: I see. I removed them.

File: tests/visualization_tests/test_timeline.py (hunk 15 above)
Comment (eukaryo) [reply tests/visualization_tests/test_timeline.py]: Thank you. I revised the code.

File: tests/visualization_tests/test_timeline.py (hunk 16 above)
Comment (eukaryo) [reply tests/visualization_tests/test_timeline.py]: I see. I changed to use `_create_study` .

File: optuna/visualization/_pareto_front.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (toshihikoyanase) [reply optuna/visualization/_pareto_front.py]: Thanks!

File: optuna/visualization/_timeline.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (toshihikoyanase) [reply optuna/visualization/_timeline.py]: Thanks!

File: tests/visualization_tests/test_timeline.py [hunk 20]
```diff
@@ -0,0 +1,61 @@
+import datetime
+from io import BytesIO
+from typing import List
```
Comment (toshihikoyanase) [inline tests/visualization_tests/test_timeline.py:3]: Could you replace `typing.List` with `list` as can be seen in https://github.com/optuna/optuna/commit/76c745be11edfa944c435b67789a2ce545615791?

We need `from __future__ import annotations` at the first line of this file.

File: tests/visualization_tests/test_timeline.py [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+                distributions={"x": optuna.distributions.FloatDistribution(-1.0, fmax)},
+                value=0.0,
+                state=s,
+            )
+        )
+    return study
+
+
+def test_get_timeline_info_empty() -> None:
+    study = _create_study([])
```
Comment (toshihikoyanase) [inline tests/visualization_tests/test_timeline.py:31]: We can directly call `optuna.create_study` here to simplify the logic.
```suggestion
    study = optuna.create_study()
```

File: tests/visualization_tests/test_timeline.py [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+        assert type(bar.hovertext) is str
+        assert isinstance(bar.start, datetime.datetime)
+        assert isinstance(bar.end, datetime.datetime)
+        assert bar.start <= bar.end
+
+
+@pytest.mark.parametrize(
+    "trial_states_list",
+    [
+        [TrialState.COMPLETE, TrialState.PRUNED, TrialState.FAIL],
```
Comment (toshihikoyanase) [inline tests/visualization_tests/test_timeline.py:53]: How about adding the empty study?
```suggestion
    [
        [],
        [TrialState.COMPLETE, TrialState.PRUNED, TrialState.FAIL],
```

**Review by toshihikoyanase — COMMENTED** [review]
Thank you for your update. Overall, it looks almost good to me, but I have some additional comments.

File: docs/source/reference/visualization/index.rst [hunk 23]
```diff
@@ -27,6 +27,7 @@ The :mod:`~optuna.visualization` module provides utility functions for plotting
    optuna.visualization.plot_param_importances
    optuna.visualization.plot_pareto_front
    optuna.visualization.plot_slice
+   optuna.visualization.plot_timeline
```
Comment (toshihikoyanase) [inline docs/source/reference/visualization/index.rst:30]: Ah, the document build failed. We need to update `optuna/visualization/__init__.py` to expose `plot_timeline` there.

```
IPython could not be loaded!
WARNING: [autosummary] failed to import optuna.visualization.plot_timeline.
Possible hints:
* ModuleNotFoundError: No module named 'optuna.visualization.plot_timeline'
* AttributeError: module 'optuna.visualization' has no attribute 'plot_timeline'
* ImportError: 
```

File: optuna/visualization/_timeline.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (HideakiImamura) [reply optuna/visualization/_timeline.py:57]: Thanks for the reply. You are right. This is necessary. By the way, I think it is useful to note that behavior in the document.

File: optuna/visualization/_utils.py (hunk 12 above)
Comment (toshihikoyanase) [reply optuna/visualization/_utils.py:194]: Ah, users may use `is_available`, but they will not use `_make_hovertext`. We should keep it private.

File: optuna/visualization/_timeline.py [hunk 24]
```diff
@@ -0,0 +1,142 @@
...
+if _imports.is_successful():
+    from optuna.visualization._plotly_imports import go
+
+_logger = get_logger(__name__)
+
+
+class _TimelineBarInfo(NamedTuple):
+    number: int
+    start: datetime.datetime
+    end: datetime.datetime
```
Comment (HideakiImamura) [inline optuna/visualization/_timeline.py:23]: It would be better to use `complete` instead of `end`.

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the update. Basically, LGTM. I have 2 minor comments. PTAL.

File: optuna/visualization/_timeline.py (hunk 24 above)
Comment (eukaryo) [reply optuna/visualization/_timeline.py]: Agreed. I revised the code.

File: optuna/visualization/_timeline.py [hunk 25]
```diff
@@ -0,0 +1,142 @@
...
+def _get_timeline_info(study: Study) -> _TimelineInfo:
+    bars = []
+    for t in study.get_trials():
+        date_complete = t.datetime_complete or datetime.datetime.now()
+        date_start = t.datetime_start or date_complete
+        if date_complete < date_start:
+            _logger.warning(
+                "In your study, trial "
+                + str(t.number)
+                + " has datetime_start later than datetime_complete."
```
[the remark this answers is not in the document]
Comment (eukaryo) [reply optuna/visualization/_timeline.py]: Thank you. I changed the message.

File: tests/visualization_tests/test_timeline.py [hunk 26]
```diff
@@ -0,0 +1,116 @@
...
+    fig = plot_timeline(study)
+    assert type(fig) is go.Figure
+    fig.write_image(BytesIO())
+
+
+def test_get_timeline_plot_negative_elapsed_time() -> None:
+    study = _create_study_negative_elapsed_time()
+    fig = plot_timeline(study)
+    assert type(fig) is go.Figure
+    fig.write_image(BytesIO())
```
Comment (toshihikoyanase) [inline tests/visualization_tests/test_timeline.py:116]: Could you remove this test case, please?
I'd like to check the rendering results of negative elapsed time in visual regression tests.

```suggestion
```

File: docs/source/reference/visualization/index.rst (hunk 23 above)
Comment (toshihikoyanase) [reply docs/source/reference/visualization/index.rst:30]: I confirmed that the reference was rendered as expected.

**Review by toshihikoyanase — APPROVED** [review]
Thank you for your update. LGTM!

**Review by HideakiImamura — APPROVED** [review]
Thanks for the update! LGTM!

**Pull request merged**